pkg/applicationserver/payload.go

```diff
@@ -193,3 +194,81 @@ func (as *ApplicationServer) decodeDownlink(ctx context.Context, dev *ttnpb.EndD
 	}
 	return nil
 }
+
+var cayenneLPPGPSKeyRegexp = regexp.MustCompile(`^gps_\d+$`)
+
+func (as *ApplicationServer) locationFromDecodedPayload(uplink *ttnpb.ApplicationUplink) (res *ttnpb.Location) {
+	fields := uplink.DecodedPayload.GetFields()
+	if len(fields) == 0 {
+		return
+	}
+
+	// Check CayenneLPP location (field gps_#).
+	for k := range fields {
+		if cayenneLPPGPSKeyRegexp.MatchString(k) {
+			if s := fields[k].GetStructValue(); s != nil {
+				res = &ttnpb.Location{
+					Latitude:  s.Fields["latitude"].GetNumberValue(),
+					Longitude: s.Fields["longitude"].GetNumberValue(),
+					Altitude:  int32(s.Fields["altitude"].GetNumberValue()),
+					Source:    ttnpb.SOURCE_GPS,
+				}
+				break
+			}
+		}
+	}
+
+	if res == nil {
+		// Check location in fields.
+		for _, pp := range []struct {
+			latKey, lonKey, altKey string
+		}{
+			{"lat", "lon", "alt"},
+			{"lat", "lng", "alt"},
+			{"lat", "long", "alt"},
+			{"latitude", "longitude", "altitude"},
+			{"Latitude", "Longitude", "Altitude"},
+			{"latitudeDeg", "longitudeDeg", "altitude"},
+			{"latitudeDeg", "longitudeDeg", "height"},
+			{"gps_lat", "gps_lng", "gps_alt"},
+			{"gps_lat", "gps_lng", "gpsalt"},
```

Comment:
If I understand this code correctly, this will only allow these 9 possible sets of keys, and no other combinations of them. Won't it be better to check one field at a time, trying all it's possibilities, before checking the next field?

Reply:
ie, say I have the fields `latitude`, `long` and `altitude`, then this algorithm won't accept it.

Reply:
Yes indeed. I wanted to make this explicit, to not trigger magic behavior when these keys aren't really consistent (like `latitudeDeg` and `gps_lng`)

Reply:
Yes I see your point. My opinion is slightly different: I try my best to obtain a valid location, no matter how "broken" the payload parser may be.

From TTN Mapper's side we recommend a specific set of keys:
https://docs.ttnmapper.org/FAQ.html#using-a-gps-tracker-device-transmitting-gps-coordinates-in-the-lora-payload
> The JSON object should contain the keys “latitude”, “longitude” and “altitude” and one of “hdop”, “accuracy” or “sats”

Something more to note here is that sats, hdop and accuracy are handled as three different things. They are inherently three different ways to indicate how reliable the gps location is. Putting hdop into a field called "accuracy" can cause confusion, as an accuracy of 3 metres is good, but an hdop of 3 is bad.